### ISSUE TITLE
in_frustum() falls off the end without a return — silently culls every Gaussian (undefined behavior, architecture-dependent)

### ISSUE THREAD

**Issue by HanSchne** [gardiens/EOGS2#6]
_Filed from outside the project team, while adapting EOGS2 for a lunar-imagery use case (SPICE/USGSCSM cameras over LRO NAC + Chandrayaan-2 TMC-2). Happy to open a PR with the one-line fix if useful._

**Repo**: [gardiens/EOGS2](https://github.com/gardiens/EOGS2)
**Affected commit**: `3cd57758ac462ce58b291e6f254d7b8a96a7f909` (`:newspaper: Add ISPRS 2026 publication details`, and every commit before it back to at least `96dbc54`)
**File**: `src/gaussiansplatting/submodules/diff-gaussian-rasterization/cuda_rasterizer/auxiliary.h`, function `in_frustum()` (lines 151–176)
**Introduced by**: `96dbc54dc6a35a798879aa7a80b3de24ef647b26` (pierrick_bournez, 2025-11-20)
**Severity**: Critical — training silently produces an empty point cloud (0 surviving Gaussians) with no error, warning, or crash.
**Found while**: reproducing the MVP1–3 lunar EOGS++ pipeline (SelenSplat project) from raw data on a fresh machine (RTX 5080 Laptop GPU, Blackwell / `sm_120`).

## Summary

`in_frustum()` in the vendored `diff-gaussian-rasterization` CUDA extension is a non-`void` function that, in its current form, **falls off the end without executing any `return` statement**. Its actual frustum-rejection logic was deliberately commented out (the comment explains `p_view.z` was repurposed to carry altitude for this satellite-imagery use case), but no `return true;` was left in its place. The missing return is undefined behavior in C++; the compiled function returns whatever happens to be left in the return-value register.

On whatever toolchain/GPU architecture this was originally built and tested against, that leftover value evidently read as truthy, so every Gaussian passed the check and training worked as intended (this is presumably why the bug has never been reported). Rebuilding the same source for `sm_120` (RTX 50-series / Blackwell) with CUDA 12.8's `nvcc` and GCC 13 as host compiler produces a binary where the same undefined behavior instead reads as **falsy** — so `preprocessCUDA()` in `forward.cu` treats every single Gaussian as outside the view frustum and culls it (`radii[idx]` stays `0`) before any rendering happens. With zero visible Gaussians, no positional or color gradient ever reaches `backward()`, and any `only_prune`-style pruning that runs on top eventually — or, as observed here, immediately — removes every point, leaving a `0`-vertex output `.ply`.

This is **not specific to any one scene, dataset, or camera model** — it reproduces identically on the project's own reference dataset (`JAX_004`, standard RPC/affine cameras) and on an entirely unrelated custom scene (lunar orbital imagery with SPICE-derived cameras), which is what ruled out a data- or camera-specific cause and pointed at the rasterizer itself.

## Reproduction

Environment: Python 3.11, PyTorch `2.10.0+cu128`, CUDA 12.8 (`nvcc`, `V12.8.61`), GCC 13 as host compiler, `TORCH_CUDA_ARCH_LIST=12.0`, RTX 5080 Laptop GPU (`sm_120`).

### 1. Reference scene (`JAX_004`, from the project's own release dataset)

```
python src/gaussiansplatting/train_pan.py \
  experiments=eogsplus mode=onlyMSI rpc_type=rpc_raw scene=JAX_004 numiterations=100 \
  model.model_path=<run_dir> run_tsdf=false run_rendering=false \
  'save_iterations=[100]' 'checkpoint_iterations=[]'
```

Result before fix: `input.ply` has 403,055 points at initialization; `point_cloud/iteration_100/point_cloud.ply` has `element vertex 0`.

### 2. Unrelated custom scene (lunar orbital imagery, LRO NAC + Chandrayaan-2 TMC-2, SPICE/USGSCSM-derived cameras)

Same symptom: 52,869 points at initialization, `0` survive after both a 100-iteration smoke run and a full 5000-iteration run (loss curve looks entirely plausible throughout — `1.04 → 0.55` — which is what made this easy to miss without inspecting the output point cloud directly).

### Diagnostic instrumentation used to localize it

Added temporarily to `train_pan.py`'s training loop, right after `render_pkg = render(cam, gaussians, pipe, bg)`:

```python
r = render_pkg["radii"]          # (required patching renderer.py to always
                                  #  populate "radii" regardless of pipe.require_radii)
print("n_visible(radii>0)", int((r > 0).sum()), "n", r.numel())
```

and right after `inter_loss.mean().backward()`:

```python
g = gaussians._opacity.grad
xyz_g = gaussians._xyz.grad
print("opacity_grad_std", float(g.std()), "opacity_grad_unique", int(torch.unique(g).numel()),
      "xyz_grad_nonzero_rows", int((xyz_g.abs().sum(dim=-1) > 0).sum()))
```

Result, identical on both scenes, every iteration, every camera:

| Quantity | Observed value | Expected value |
|---|---|---|
| `n_visible(radii>0)` | **0** (out of 403,055 / 52,869) | > 0 |
| `gaussians._xyz.grad` | exactly `0` for every point | per-point, geometry-dependent |
| `gaussians._opacity.grad` | a single value broadcast to every point (`std == 0`, `unique == 1`) | per-point, visibility-dependent |

The uniform, nonzero `_opacity.grad` is a red herring — it comes entirely from a pure-PyTorch opacity regularizer (`OpacityLoss`/`AccumulatedOpacity`) that never touches the CUDA kernel; since all opacities start at the identical initial value, its analytic gradient is naturally identical across all of them too. The real signal is `n_visible(radii>0) == 0`: **nothing is ever rasterized**, for any camera, in any scene.

## Root cause

`cuda_rasterizer/forward.cu`, `preprocessCUDA()`:

```cpp
radii[idx] = 0;
...
if (!in_frustum(idx, orig_points, viewmatrix, projmatrix, prefiltered, p_view))
    return;
```

`cuda_rasterizer/auxiliary.h`:

```cpp
__forceinline__ __device__ bool in_frustum(int idx,
    const float* orig_points,
    const float* viewmatrix,
    const float* projmatrix,
    bool prefiltered,
    float3& p_view)
{
    float3 p_orig = { orig_points[3 * idx], orig_points[3 * idx + 1], orig_points[3 * idx + 2] };

    // Bring points to screen space
        // float4 p_hom = transformPoint4x4(p_orig, projmatrix);
    // float p_w = 1.0f / (p_hom.w + 0.0000001f);
    // float3 p_proj = { p_hom.x * p_w, p_hom.y * p_w, p_hom.z * p_w };

    p_view = transformPoint4x3(p_orig, viewmatrix);
    // This is commented out because p_view.z contains the altitude
    // if (p_view.z <= 0.2f)// || ((p_proj.x < -1.3 || p_proj.x > 1.3 || p_proj.y < -1.3 || p_proj.y > 1.3)))
    // {
    // 	if (prefiltered)
    // 	{
    // 		printf("Point is filtered although prefiltered is set. This shouldn't happen!");
    // 		__trap();
    // 	}
    // 	return false;
    // }
}   // <-- falls off the end; no `return` statement anywhere in this function body
```

`git blame` confirms this exact state (comment block included) has been in place since the function's introduction in `96dbc54` — it isn't a copy/build artifact or something broken by an intermediate edit, the `return` was simply never added when the frustum check was disabled.

A `bool`-returning function with no `return` on a reachable path is undefined behavior per the C++ standard; `nvcc`/`g++` are not obligated to warn (and did not, even at `-Wall`, since `__forceinline__ __device__` functions are inlined into a much larger caller before most such diagnostics would fire). In practice the "returned" value is whatever bit pattern is sitting in the register the calling convention uses for a `bool` return — which is a function of the surrounding generated code, and therefore of the exact compiler version, optimization level, and target architecture. That is precisely the kind of thing a from-source rebuild for a new GPU architecture (`sm_120`, first-generation Blackwell) with a different `nvcc`/host-compiler combination can flip, with zero change to the actual source.

## Fix

```diff
 	p_view = transformPoint4x3(p_orig, viewmatrix);
 	// This is commented out because p_view.z contains the altitude
 	// if (p_view.z <= 0.2f)// || ((p_proj.x < -1.3 || p_proj.x > 1.3 || p_proj.y < -1.3 || p_proj.y > 1.3)))
 	// {
 	// 	if (prefiltered)
 	// 	{
 	// 		printf("Point is filtered although prefiltered is set. This shouldn't happen!");
 	// 		__trap();
 	// 	}
 	// 	return false;
 	// }
+	return true;
 }
```

This restores the evident original intent (frustum rejection deliberately disabled → every point should pass) instead of leaving the outcome to undefined behavior.

## Verification

Rebuilt `diff-gaussian-rasterization` for `sm_120` with the one-line fix applied (`pip install --no-build-isolation` with `TORCH_CUDA_ARCH_LIST=12.0`, conda CUDA 12.8 `nvcc`, GCC 13 as host compiler), then re-ran both 100-iteration smoke tests from a clean point cloud:

| Scene | Points at init | Points surviving, before fix | Points surviving, after fix |
|---|---|---|---|
| `JAX_004` (reference dataset) | 403,055 | 0 | **281,492** |
| Lunar (`SARABHAI_CAMERA_CORE`) | 52,869 | 0 | **52,869** |

Both scenes now produce non-empty, actively-training point clouds.

The lunar scene's full 5000-iteration run was also re-executed end-to-end with the fix in place, and compared directly against the previous full run under the buggy binary (same scene, same config, same seed):

| Metric (iteration 5000) | Before fix | After fix |
|---|---|---|
| Surviving Gaussians | **0** | **17,220** (of 52,869 initial) |
| Total training loss | 0.53 | **0.23** |
| Train-view PAN PSNR | 13.4 dB (plateaus by ~iter 400, from color-correction params alone — no real 3D structure) | **19.1 dB** (still climbing at iteration 5000) |
| Train-view PAN SSIM | 0.27 | **0.47** |

The buggy run's loss and PSNR curves look entirely plausible in isolation — they decrease/improve smoothly and plateau at a "reasonable-looking" value — which is exactly why this bug is easy to miss without directly inspecting the output point cloud's vertex count. The fixed run is not only non-empty but measurably better on every reconstruction-quality metric, since real per-Gaussian gradients are now reaching the geometry instead of the model relying entirely on non-geometric parameters (color correction, exposure) to fit the training views.

## Suggested upstream action

- Apply the one-line fix above to `cuda_rasterizer/auxiliary.h`.
- Consider adding `-Wreturn-type` as a hard error for this extension's build (it already fires for host C++ via `-Wall`, but device-side `__forceinline__` functions inlined into a template-heavy caller can suppress the diagnostic; an explicit check would catch a regression like this immediately instead of it silently depending on undefined behavior for years).
- Consider asserting `radii.sum() > 0` (or logging the visible-Gaussian fraction) somewhere in the training loop as a cheap sanity check — this class of bug produces no error and a perfectly plausible-looking loss curve, so the only way to notice it is to inspect the output point cloud directly.

**Comment by gardiens** [comment]
Hello , 
What do you suggest in the end ? 
Adding the return true in the rasterizer?

**Comment by gardiens** [comment]
Thanks @HanSchne, this is an excellent bug report, exactly the kind of clear before/after reproduction that makes a fix trivial to validate.

Applying return true; in in_frustum() as suggested. This restores the original intent, since the frustum check was deliberately disabled and every point is meant to pass. The before/after numbers on both scenes confirm the fix.


Closing this out now that the fix is merged. Thanks again for the thorough writeup:)
